### ISSUE TITLE
Empty select() should not project asterisk if unknown table is used with leftSemiJoin() or leftAntiJoin()

### ISSUE THREAD

**Issue by kaini** [jOOQ/jOOQ#12328]
### Expected behavior

When using a `leftSemiJoin` with an *unknown* table (and I guess - but did not verify - when using anti-joins as well) jOOQ throws away the column information that is used to map the result set into records. This leads to an ambiguous/invalid mapping since `into` falls back to trying to maching column and field names.

This is to be expected to happen for all join types that change the projection, since there is no way for jOOQ to know the columns. But semi- and anti-joins do not change the projection, therefore I believe this to be a bug.

```
        YRecord wrong = ctx
			.select()
			.from(X)
			.join(Y)
			.on(Y.ID.eq(X.ID))
			.leftSemiJoin(DSL.table(DSL.name("OTHER")))
			.on(DSL.field(DSL.name("OTHER", "ID"), Long.class).eq(Y.VALUE))
			.fetchOne()
			.into(Y);
```

In this case the `.into(Y)` uses columns from `X` if they have the same name as columns in `Y`. (See MCVE)

```
		YRecord right = ctx
			.select()
			.from(X)
			.join(Y)
			.on(Y.ID.eq(X.ID))
			.where(DSL.exists(DSL.select(DSL.one()).from(DSL.table(DSL.name("OTHER"))).where(DSL.field(DSL.name("OTHER", "ID"), Long.class).eq(Y.VALUE))))
			.fetchOne()
			.into(Y);
```

This query works, because jOOQ keeps the information about which column is which.

### Steps to reproduce the problem

You can find an MCVE here: https://github.com/kaini/jOOQ-mcve-1

### Versions

- jOOQ: 3.15.1 Enterprise
- Java: 16
- Database (include vendor): Oracle 19c (also observed with DB2 LUW 11.5)
- OS: Linux
- JDBC Driver (include name if inofficial driver): `com.oracle.database.jdbc:ojdbc11:21.1.0.0` (also observed with `com.ibm.db2:jcc:11.5.5.0`)

**Comment by lukaseder** [comment]
Thanks a lot for your report.

I agree the synthetic semi/anti join operators shouldn't be taken into consideration when assessing whether we know the auto-projection or not.

Of course, the workaround is always to be explicit about the projection, not just to avoid this issue, but also to make for more optimal queries. In your specific example, you shouldn't project any columns of `X` because you don't need them:

```java
        YRecord works = ctx
			.select(Y.fields())
			.from(X)
			.join(Y)
			.on(Y.ID.eq(X.ID))
			.leftSemiJoin(DSL.table(DSL.name("OTHER")))
			.on(DSL.field(DSL.name("OTHER", "ID"), Long.class).eq(Y.VALUE))
			.fetchOne()
			.into(Y);
```

Alternatively, if you need `X` and `Y` in your real world example, the workaround is to project both:

```java
        YRecord works = ctx
			.select(X.fields())
			.select(Y.fields())
			.from(X)
			.join(Y)
			.on(Y.ID.eq(X.ID))
			.leftSemiJoin(DSL.table(DSL.name("OTHER")))
			.on(DSL.field(DSL.name("OTHER", "ID"), Long.class).eq(Y.VALUE))
			.fetchOne()
			.into(Y);
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0 and 3.15.2 (https://github.com/jOOQ/jOOQ/issues/12329). Thanks again for your report!
